tests/utils/log_reader.py

```diff
@@ -27,77 +28,109 @@
-    """Reader that allows to read file in separate thread and read and wait for new lines"""
+    """Reader that allows to read file and wait for new lines"""
 
     def __init__(self, file_path):
         super(LogReader, self).__init__()
-        self.daemon = True
+        self._file_path = file_path
         self._file = open(six.text_type(file_path), "r")
-        self._stop_event = threading.Event()
-        self._lines_lock = threading.Lock()
-        self._start_cv = threading.Condition()
-        self._has_data = False
-        self._lines = collections.deque(maxlen=1000)
-        self._wait_for_line_cv = threading.Condition()
-
-        self._has_new_line = False
-
-    def handle_new_line(self, line):
-        with self._lines_lock:
-            self._lines.append(line)
-        with self._wait_for_line_cv:
-            self._has_new_line = True
-            self._wait_for_line_cv.notify()
-
-    def run(self):
-        def get_line():
-            while not self._stop_event.is_set():
-                _line = self._file.readline()
-                if _line:
-                    return _line
-                time.sleep(0.1)
-
-        get_line()
-        with self._start_cv:
-            self._has_data = True
-            self._start_cv.notify()
-        self._file.seek(0)
-        while not self._stop_event.is_set():
-            line = get_line()
-            self.handle_new_line(line)
-
-    def start(self, wait_for_data=False):
-        super(LogReader, self).start()
-        if not wait_for_data:
-            return
-        with self._start_cv:
-            while not self._has_data:
-                self._start_cv.wait(0.1)
-        return
+        self._error_line_patterns = dict()
+        self._lines = list()
+
+    def _check_line_for_error(self, line):
+        for pattern, (compiled_pattern, message) in self._error_line_patterns.items():
+            if compiled_pattern.match(line):
+                if not message:
+                    message = "Log file {0} contains error line: '{1}'. Pattern: {2}.".format(
+                        self._file_path, line, pattern
+                    )
+                raise LogReaderError(message)
+
+    def _handle_new_line(self, line):
+        self._lines.append(line)
+        self._check_line_for_error(line)
+
+    def _read_line(self):
+        line = self._file.readline()
+        if line:
+            self._handle_new_line(line)
+
+        return line
+
+    def wait_for_next_line(self, timeout=10):
+        timeout_time = time.time() + timeout
+        while True:
+            line = self._read_line()
+            if line:
+                return line
+
+            if time.time() >= timeout_time:
+                raise LogReaderTimeoutError(
+                    "Timeout of %s seconds reached while waiting for metrics" % timeout
+                )
+
+            time.sleep(0.01)
+
+    def go_to_end(self):
+        while True:
+            line = self._read_line()
```

Comment:
IIRC, in some scenarios, underlying ``readline()`` will block until there is data available / EOF is reached, right?

Reply:
readline() does not block if I am not mistaken. It just returns an empty string.